### PR TITLE
saving indicator for notes

### PR DESCRIPTION
**Pull request by benq66**

### CODE REVIEW

File: src/components/Dialog-Window/Notes/DialogNote.tsx [hunk 1]
```diff
@@ -10,14 +10,38 @@ export type NoteProps = {
 export const Note: React.FC<NoteProps> = ({ maxLength, id }) => {
   const [userData, setUserData] = useLocalStorage(id);
   const [note, setNote] = React.useState(userData[id].note ?? "");
+  const [dynamicSavingText, setDynamicSavingText] = React.useState("");
+  const [savingTextTimeout, setSavingTextTimeout] =
+    React.useState<NodeJS.Timeout>();
+
+  // TODO: Translate
+  const savingTextLabel = "lagring...";
+  const savedTextLabel = "lagret";
```
Comment (boyum) [inline src/components/Dialog-Window/Notes/DialogNote.tsx:19]:
```suggestion
  const savingTextLabel = "Saving...";
  const savedTextLabel = "Saved";
```

File: src/components/Dialog-Window/Notes/DialogNote.tsx [hunk 2]
```diff
@@ -10,14 +10,38 @@ export type NoteProps = {
...
+
+  const setSavingText = (): void => {
+    setDynamicSavingText(savingTextLabel);
+    setSavingTextTimeout(
+      setTimeout(() => {
```
Comment (boyum) [inline src/components/Dialog-Window/Notes/DialogNote.tsx:24]: `setTimeout` looks to tsc like it returns a NodeJS.Timeout, but is in fact `window.setTimeout`, which returns a `number`.

This won't change anything, but might clear up the typing for future devs.
 
```suggestion
      window.setTimeout(() => {
```

File: src/components/Dialog-Window/Notes/DialogNote.tsx [hunk 3]
```diff
@@ -10,14 +10,38 @@ export type NoteProps = {
...
+  };
 
   React.useEffect(() => {
     // TODO: If this becomes laggy, add a debounce-timer to avoid saving more often than, say, every 100ms.
     userData[id].note = note;
     setUserData(userData);
-  }, [userData, id, note, setUserData]);
+    // ensure there's no memory leak on component unmount during timeout
+    return () => {
+      if (savingTextTimeout) clearTimeout(savingTextTimeout);
```
Comment (boyum) [inline src/components/Dialog-Window/Notes/DialogNote.tsx:39]: There's is a minor chance that `savingTextTimeout` might be `0`, which is falsy. It's better to check for nullish values.

```suggestion
      if (savingTextTimeout != null) clearTimeout(savingTextTimeout);
```

File: src/components/Dialog-Window/Notes/DialogNote.tsx [hunk 4]
```diff
@@ -10,14 +10,38 @@ export type NoteProps = {
...
+  const [savingTextTimeout, setSavingTextTimeout] =
+    React.useState<NodeJS.Timeout>();
+
+  // TODO: Translate
+  const savingTextLabel = "lagring...";
+  const savedTextLabel = "lagret";
+
+  const setSavingText = (): void => {
+    setDynamicSavingText(savingTextLabel);
+    setSavingTextTimeout(
```
Comment (boyum) [inline src/components/Dialog-Window/Notes/DialogNote.tsx:23]: Should we perhaps clear any existing timeouts before starting a new one?

File: src/components/Dialog-Window/Notes/DialogNote.tsx (hunk 3 above)
Comment (boyum) [reply src/components/Dialog-Window/Notes/DialogNote.tsx:41]: It's really good that we're clearing timeouts before unmount, though! Great catch 🎉

**Comment by boyum** [comment]
Looks great!!!!!

File: src/components/Dialog-Window/Notes/DialogNote.tsx (hunk 4 above)
Comment (benq66) [reply src/components/Dialog-Window/Notes/DialogNote.tsx:23]: This should be already achieved by the clearTimeout in useEffect as far as I understand. At least at the moment the timer gets reset each time onChange is called.

File: src/components/Dialog-Window/Notes/DialogNote.module.scss [hunk 5]
```diff
@@ -13,3 +13,10 @@
   bottom: -1.5rem;
   right: -0.3rem;
 }
+
+.dynamicSavingText {
+  color: gray;
```
Comment (boyum) [inline src/components/Dialog-Window/Notes/DialogNote.module.scss:18]: The contrast with `gray` foreground on `white` background is too low. Perhaps we can use one of the theme colors here?

```suggestion
  color: var(--theme-color-2);
```

Comment (boyum) [reply src/components/Dialog-Window/Notes/DialogNote.module.scss:18]: I checked `--theme-color-2` for our three themes, and the contrast value is good enough on all of them.

Comment (boyum) [reply src/components/Dialog-Window/Notes/DialogNote.module.scss:18]: Let's forget about this now and do it in an upcoming PR.

Comment (boyum) [reply src/components/Dialog-Window/Notes/DialogNote.module.scss:18]: https://github.com/NDLA-H5P/h5p-topic-map/pull/184

File: src/components/Dialog-Window/Notes/DialogNote.tsx (hunk 4 above)
Comment (boyum) [reply src/components/Dialog-Window/Notes/DialogNote.tsx:23]: Ok!

File: src/components/Dialog-Window/Notes/DialogNote.tsx (hunk 1 above)
Comment (benq66) [reply src/components/Dialog-Window/Notes/DialogNote.tsx:19]: ✅

File: src/components/Dialog-Window/Notes/DialogNote.tsx (hunk 2 above)
Comment (benq66) [reply src/components/Dialog-Window/Notes/DialogNote.tsx:24]: ✅

File: src/components/Dialog-Window/Notes/DialogNote.tsx (hunk 3 above)
Comment (benq66) [reply src/components/Dialog-Window/Notes/DialogNote.tsx:41]: ✅

File: src/components/Dialog-Window/Notes/DialogNote.module.scss (hunk 5 above)
Comment (benq66) [reply src/components/Dialog-Window/Notes/DialogNote.module.scss:18]: Absolutely, looks much better!

**Review by boyum — APPROVED** [review]

**Pull request merged**